Gopkg.toml

```diff
@@ -83,6 +91,10 @@ required = [
   branch = "master"
   name = "github.com/giantswarm/micrologger"
 
+[[constraint]]
+  branch = "master"
+  name = "github.com/giantswarm/microstorage"
```

Comment:
I am wondering why we need storage now when there is a controller that manages the network ranges. Is it just used for intermediate caching? I think when we use the storage for more than caching the implementation of the operator here is wrong and not necessary at all because we then could also have kept the dumb kubernetesd approach. My thinking here was that we rely on the desired state from the CR and the current state from the systems we work with, and that is it. So we should then also have different IPAM resource implementations for Azure and AWS because we read VPCs and whatever equivalent we leverage on Azure. This would then be a real step forward and the system would be way more able to heal itself as opposed to relying on some decoupled storage data.

Reply:
Reason for using storage here is to proceed step by step. My mid-far future plan here is to remove persistence for allocated subnets from microstorage and start using existing `ClusterNetworkConfig` CRs for that, but before all provider specific resources are implemented and current network allocations are not migrated, this is my first step. When getting provider specific resources done and migrating those from `kubernetesd` to `cluster-operator`, existing subnet allocations can be migrated with `ipam.Migrator`.

As I tried to explain earlier, existing ipam implementation cannot be used with CRUD resources (and we need CRUD resources for provider specific resource implementations) because subnet must be allocated in `GetDesiredState()` which is independent from `GetCurrentState()` and these are again separated from `ApplyCreateChange()` where decision on whether or not desired state should be created exists. There is plenty of room for various error situations due to runtime errors or plain simple bugs that interrupt the single resource reconciliation execution before getting to the point of cleanup.

An example: Guest cluster network is `10.1.0.0/16` and `cidrMask` for single guest cluster subnet is `24`. This means there are 255 available subnets for guest clusters to use. If due to an error (could be a very simple bug in some specific corner case) subnets are allocated but not released nor used, with 5min informer refresh time all subnets are allocated in 21h and after this no new guests can be created any more. Relying on specific ordering in resource execution or current implementation in OperatorKit is not robust design IMO when there is an alternative of encapsulating specific logic to one place that can safely manage it. That's why I created specific CRD and specific resource for managing IPAM of guest cluster networks.

Regarding different providers in current or foreseeable future affect how our platform shares guest cluster subnets. To support Azure limitations, there is notion of reserved subnets, but that implementation detail makes sense also in other environments.

I think @kopiczko and @JosephSalisbury have earlier worked on IPAM. Maybe you have some thoughts about this?

Reply:
I have a thought. Maybe it doesn't need to be so complex.
At some point there was `annotation` arg added to the `CreateSubnet` method: https://github.com/giantswarm/ipam/blob/9f5e82f904b3f54b19ac8cb58cbbadb59d50eb89/service.go#L104

We could add `FindSubnet(annotation string)` method which would allow to determine if the subnet was allocated for given cluster ID. Annotation is just a cluster ID. For new guest clusters storage looks like:

```yaml
# Output of the command below on gollum (plus some comments):
# $ k get -n giantswarm storageconfig kubernetesd -o yaml

apiVersion: core.giantswarm.io/v1alpha1
kind: StorageConfig
metadata:
  clusterName: ""
  creationTimestamp: 2017-12-15T11:29:32Z
  generation: 0
  name: kubernetesd
  namespace: giantswarm
  resourceVersion: "33598445"
  selfLink: /apis/core.giantswarm.io/v1alpha1/namespaces/giantswarm/storageconfigs/kubernetesd
  uid: 385e9ac6-e18b-11e7-8c19-000d3a2c9706
spec:
  storage:
    data:
      /ipam/subnet/10.0.0.0-16: 10.0.0.0/16
      /ipam/subnet/10.1.0.0-16: 3p5j2
      /ipam/subnet/10.2.0.0-16: 10.2.0.0/16
      /ipam/subnet/10.3.0.0-16: 10.3.0.0/16
      /ipam/subnet/10.4.0.0-16: 10.4.0.0/16
      /ipam/subnet/10.5.0.0-16: 10.5.0.0/16
      /ipam/subnet/10.6.0.0-16: 10.6.0.0/16
      /ipam/subnet/10.7.0.0-16: 10.7.0.0/16
      /ipam/subnet/10.8.0.0-16: 10.8.0.0/16
      /ipam/subnet/10.9.0.0-16: 10.9.0.0/16
      /ipam/subnet/10.10.0.0-16: 10.10.0.0/16
      /ipam/subnet/10.11.0.0-16: 10.11.0.0/16
      /ipam/subnet/10.12.0.0-16: 10.12.0.0/16  # For an older cluster prior to annotation.
      /ipam/subnet/10.13.0.0-16: ep5sx         # For a newer cluster after annotation were intoruced.
      /ipam/subnet/10.14.0.0-16: 6hgho
      /ipam/subnet/10.15.0.0-16: f4bna
      /ipam/subnet/10.16.0.0-16: 27v9c
```

If we could migrate older cluster and introduce `FindSubnet` wouldn't it be simpler to implement? Would there be any corner case that can't be covered?

Reply:
> We could add FindSubnet(annotation string) method which would allow to determine if the subnet was allocated for given cluster ID.

This sounds like a good idea for replacing current "delete subnet allocation if UpdateStatus fails" logic. This way the resource can reconcile on error and there wouldn't be orphaned subnet entries in storage. When replacing microstorage with logic that maintains state purely with present CR objects, then this is not needed anymore, but for intermediate solution, it would improve robustness.

Reply:
> When replacing microstorage with logic that maintains state purely with present CR objects, then this is not needed anymore, but for intermediate solution, it would improve robustness.

Or as we discussed other way around when `FindSubent` would be in place ` replacing microstorage with logic that maintains state purely with present CR objects` wouldn't be needed anymore.

Reply:
I feel like I talk against a wall here and stop bothering now. How the allocation is treated here is just wrong but I give up now.

Reply:
> I feel like I talk against a wall here and stop bothering now. How the allocation is treated here is just wrong but I give up now.

Please don't give up. I would like to understand your point, but I think I don't get it right. Personally I see problem with independent execution steps in CRUD resource, but do you mean something else than what I'm picturing here.

Can you briefly describe sequence of subnet allocation and responsibility of different components and functions in the design that you see best?

Reply:
Super simplified. Consider the complete available range on a host cluster of `0-100`. There is no more space. We have to deal with it. The same is with the ingress ports and VNIs. For instance on `centaur` we have ports registered in the installation's LB to create 9 clusters or something. There is no more space. Back to IPAM.

- Guest cluster 1 has range `0-4`.
- Guest cluster 2 has range `5-10`.
- Guest cluster 3 has range `11-14`.
- Guest cluster 2 gets deleted so range `5-10` is freed eventually. 
- Guest cluster 4 should now get range `15-20`.
- Guest cluster 4 creation fails somehow but reconciliation continues.
- Guest cluster 4 should now get range `15-20` on the next reconciliation.

I hope this explains how it should be IMO. When all ranges are used we cannot create clusters anymore and this problem is not new. This is the same algorithm as implemented in `rangepool` and again IMO there should just be an abstraction to manage IPs instead of simple numbers because the algorithm is already there and just works. We can leverage what we have. 

And yeah, not to mix topics but as I said the microstorage implementation is also wrong but I agree that this is reasonable as an intermediate step. At the end the design should be to rely on the provider state and not on some DB state we have to maintain. Because this will just diverge when shit hits the fan.

Reply:
What you described there, is effectively the algorithm of `ipam.Free()`. I'm not changing that in here or even suggesting to be changed right now. It's task to be conducted in https://github.com/giantswarm/giantswarm/issues/3352 and I'd like to see allocation algorithm to be completely separated from the implementation of maintaining state of allocated and free network space. It should just receive state and calculate based on that.

I think here the relevant part you described is this:
> And yeah, not to mix topics but as I said the microstorage implementation is also wrong but I agree that this is reasonable as an intermediate step. At the end the design should be to rely on the provider state and not on some DB state we have to maintain. Because this will just diverge when shit hits the fan.

I think here we agree that microstorage is just intermediate step to get things going and be able to move forward with provider specific resource implementations.

Last bit regarding use of provider state instead of DB state I don't understand. `cluster-operator` doesn't use provider APIs as that is responsibility of provider specific operator and besides of that, AFAIK provider specific operators use given subnets as-is which leads to situation that in no circumstances cloud environment should not diverge from what is the state understood by `cluster-operator`.

Do you suggest totally different approach towards IPAM implemented roughly as follows:

- Keep IPAM algorithm separate, but changed towards rotation in order to fix ordering issues.
- Maintain state of allocated subnets in provider specific CR objects (example for AWS):
  - In `AWSConfig` resource's `GetDesiredState()`: 
    1. List `AWSConfig` CR objects.
    2. For each CR object, select subnet.
    3. Given existing subnets, "IPAM, find me a next free subnet".
    4. Store this subnet in desired state.
  - If above desired state gets created, then given subnet is allocated.
- The same for ingress ports and Flannel VNIs (KVM).

Is that a design that you see?

Reply:
> What you described there, is effectively the algorithm of `ipam.Free()`. 

No it is not. There is no rotation and the implementation is separated and different which is why we have things like https://github.com/giantswarm/giantswarm/issues/3352. 

> Is that a design that you see?

No. My idea is to have an `awsipam` resource that checks if the guest cluster has already a network range assigned by looking up VPCs or whatever we need to do against the AWS API. When it is there we read it and are done. When we do not find anything we allocate via the magic we have in the IPAM lib. When we have to delete we ensure freeing the allocated network range only once the VPC or whatever resources we rely on in AWS got removed. Because only when the resource leveraging the allocated network range does not use it anymore we have full control over preventing conflicts for real because we leverage the current state of the actual system and not any indirect that promises something and fails eventually with that promise. Then same with Azure just that we work against the current state of the Azure API.

Reply:
> My idea is to have an awsipam resource that checks if the guest cluster has already a network range assigned by looking up VPCs or whatever we need to do against the AWS API. 

Can't there be a race like: creating the stack is in progress, we check the VPCs in the meantime and there is a potential collision?

Reply:
I don't see what should go wrong. There is no race of two entities doing the same thing. So I do not follow your concerns here.

Reply:
Say we have VPCs using networks 10.0.1.0/24 and 10.0.2.0/24.

Then there are two cluster creation requests roughly at the same time: cluster A and B.

aws-operator sees CR for cluster A and checks allocated networks and allocates next free one which is 10.0.3.0/24 for cluster A. It creates a CF stack, and cancels reconciliation.

Then aws-operator sees CR for cluster B and checks allocated networks. CF for cluster A hasn't completed yet, no new VPC is created, so 10.0.3.0/24 isn't allocated and aws-operator again allocates 10.0.3.0/24 for cluster B.

Isn't that possible?

Reply:
> aws-operator sees CR for cluster A and checks allocated networks and allocates next free one which is 10.0.3.0/24 for cluster A

No, `aws-operator` doesn't do anything to "allocate" the network range. This is the job of `cluster-operator`. It figures out which range a cluster should use. This is the single entity being responsible to figure that out and the result is just given to the `AWSConfig` CR. Then the `aws-operator` acts upon this and all is fine.

Reply:
Oh yeah. I think I see where you are pointing at. The solution to what you describe would be to check if a VPC is there and if a `AWSConfig` is there and compute the desired result that way. So when either is there it means the network range is allocated. Only of there is no VPC and no CR the range is freed.

Reply:
You are right I confused things a bit. But in that case the described situation seems to be even worse. Let's say aws-operator is crashlooping so basically does nothing.

Then two clusters A and B are requested while aws-operator is still down.

Then cluster-operator lists VPCs twice (with the same output) and creates two awsconfigs with the same range based on listed VPCs.

Am I missing something?

Reply:
> The solution to what you describe would be to check if a VPC is there and if a AWSConfig is there

I'm more concerned about allocation in this example. But when we check AWSConfigs this is fine. But I believe checking also VPCs is redundant, because AWSConfig should be present as long as related VPC is still there. So it should be sufficient to only check existing AWSConfigs. Then the solution is generic. And it boils down to deallocating at the right time, not to the underlying storage we use for tracking allocated networks. It's the same thing I described here https://github.com/giantswarm/giantswarm/issues/3352#issuecomment-398747222

Reply:
> But I believe checking also VPCs is redundant, because AWSConfig should be present as long as related VPC is still there.

I don't think this is redundant. The VPC can still be there even if there is no CR. We had this already. We can only ensure allocations being freed reliably when the VPC which is actually making use of the allocation is removed.

Reply:
> The VPC can still be there even if there is no CR.

Isn't that a bug? Shouldn't we keep the CR as long as there exists any resource created as consequence of the CR?

Reply:
I can just go to the cluster and delete the CR if I want. Etcd data can just get corrupted and the backup is not fresh enough and we replay and state diverges. Then you have to go to fix it manually.

Reply:
There are many ways to break things and make them unrecoverable w/o manual intervention or unrecoverable at all. Question is if this is enough of a reason to mix logic and encapsulated knowledge of different operators here. This is fishy IMO.

Reply:
> mix logic and encapsulated knowledge of different operators

I don't follow here. What is mixed or wrong in your opinion? I see the `cluster-operator` as being mixed already. That is the actual flaw. We should actually run a different `api`, `cluster-service`, `cluster-operator` and what not per provider to not mix things up. We have the `cluster-operator` implementing different provider logic because of practicality. It making use of the provider APIs is totally reasonable for me. When you want to fix something there we have to have a dedicated `cluster-operator` per provider.

Reply:
We have separated provider logic for creating CRs. CRs differ between specific operators and that is expected. The only knowledge is has is how to create the CR, but it doesn't know how the operator behaves internally. Correct me if I'm wrong but AFAIK cluster-operator doesn't access specific provider API at the moment. If it had it would have to have knowledge on specific operator logic which isn't the case right now. E.g. knowledge on how aws-operator creates/deletes VPCs.

Reply:
I see what you mean. IMO it is not a bad thing to make the `cluster-operator` aware of VPCs. This is the bit that makes the system more reliable. I understand that this is not the most pretty thing to do but for me the system reliability counts and I do not see that the design is awfully bad when we make the `cluster-operator` aware of VPCs. IMO the system we build requires `cluster-operator` to be aware of the provider specific aspects. Hiding these details is not an improvement. We have to deal with provider specifics. We do not gain anything here when we try to only let the `aws-operator` deal with AWS. When we want to achieve that we can throw `cluster-operator` away and implement everything in the provider operators. We are not doing this though to have the benefit of having most of the common ground of cluster creation managed by `cluster-operator`. So this is the tradeoff here.

Reply:
> When IPAM with provider APIs would be the solution, I think provider specific operator should implement this, especially when there's specialized implementation for each provider (out of natural dependency to differentiating APIs).

When _it is_ a solution, I agree it should be implemented in provider specific implementation.

But I still believe we have only two problems:

- Dealing with errors during allocation. This can be solved with `FindSubnet(annotation)` method I proposed earlier.
- Deallocating network at the right time. Here I think it's enough to do two things. First thing is to not deallocate network as long as AWSConfig still exists for the cluster. Even if there is deletion timestamp set. Second thing is making sure AWSConfig is not deleted as long as _any_ resource created for that AWSConfig still exists (aws-operator responsibility). The latter isn't trivial but it should be a principle we rely on to make the designs simpler. This is how operator should work IMO - no dangling resources, as deletion must be reconciled.

These two problems are rather easy to cope with from the cluster-operator perspective. I still don't see why we need to deal with this complex design. I feel it can introduce even more problems we can't foresee yet. Is it just because you don't like crdstorage? What's wrong with it?

Reply:
- `FindSubnet(annotation)` would simplify current logic a bit and I like the idea.

- Connecting presence of `AWSConfig` to deallocation of network would require maintenance of inter-resource state which feels a bit fishy to me. I'd like to rather fix this problem with rotating allocation algorithm as Tim has suggested.

Reply:
...and regarding `crdstorage`, I don't like it that much, but I think it can be replaced with functionality operating on present CRs. That should work all fine and this is similar to what is planned for `cluster-service` as well (removing cluster object and querying present `ClusterConfig` CRs).

Reply:
I agree it can be replaced that by many other things. I'm just asking why it should be replaced?

Reply:
Use of CR object as persistent storage feels wrong to me. If we need external persistent storage, I'd rather use etcd for that (since it's already backed up anyway).

Reply:
At the end of the day you want to use multiple CRs as a persistent storage. So it's basically the same thing but it makes things more difficult to maintain and understand IMO. Also CR contents land in the etcd anyway. We were in the place where we had many etcds and it was a huge pain to maintain. That's the reason why crdstorage (previous tprstorage) emerged.

Reply:
Having said that I still don't understand why replacing crdstorage with multiple CRs is beneficial.  I can only disagree and commit.

Reply:
Granted that you do have a point here.

Reply:
> I also agree that cluster-operator should not touch provider APIs when there is provider specific operator doing this responsibility.

I strongly disagree. We have provider support in the `cluster-operator`. Not implementing provider specifics there is basically ignoring the consequence of the design we implemented. When you want to keep `cluster-operator` provider independent all the provider specifics we already have in there must be removed. I am cool with either way but you are contracting the reality here now. Either we implement the next generation of IPAM in a reliable way, which implies leveraging provider specifics like the AWS or Azure API, or we strip off everything provider specific from `cluster-operator` and find a "better" place for the functionality we have to support. Again, I do not mind either way. We just put a lot of common ground into `cluster-operator` for convenience. I want to remind about Timo's opinion that we should have created different projects for certain things long time ago. When you are now dogmatic about this we have to have an "AWS IPAM" operator to sort the network range allocation. Whatever. The point that is very important to me is that we have to respect the current state of the system we manage. This includes considering VPCs on AWS. Not doing this is totally wrong.

Reply:
Correct me if I'm wrong, but so far cluster-operator has only knowledge on how to construct the CR and this is a clear border. Having a border is not dogmatic IMO. It's practical to know where responsibilities of the project end.

Reply:
Grep the project for provider and tell me where the line is. When this does not already smell I cannot help here. What I hear from your argumentation is that we are going to ignore likely failure scenarios because we do not work against the cloud provider APIs. I do not accept that. because it is wrong and I did not hear a reasonable argument against that. All I hear is that fixing problems would not fit into the project design. What do we have to do then? Change the project design.

Reply:
I grepped the project. I still think cluster-operator just knows how to create CRs for specific operators and that's all I see.

I quote what I wrote here: https://github.com/giantswarm/cluster-operator/pull/153#discussion_r197024894

>But I still believe we have only two problems:
>
>- Dealing with errors during allocation. This can be solved with FindSubnet(annotation) method I proposed earlier.
>- Deallocating network at the right time. Here I think it's enough to do two things. First thing is to not deallocate network as long as AWSConfig still exists for the cluster. Even if there is deletion timestamp set. Second thing is making sure AWSConfig is not deleted as long as any resource created for that AWSConfig still exists (aws-operator responsibility). The latter isn't trivial but it should be a principle we rely on to make the designs simpler. This is how operator should work IMO - no dangling resources, as deletion must be reconciled.

I think checking provider for existing VPCs is just working around aws-operator bug. So it isn't a design problem IMO.

I strongly believe aws-operator should be responsible for garbage collecting resources it creates and _we should rely_ on that.

Reply:
`aws-operator` is not allocating the network range. When you want it to do it fine with me. The important point I am not letting go is that the actual state of the system has to be considered as basis of the current state. Redirecting responsibility of allocated networks being defined by the `AWSConfig` CR is wrong. The `AWSConfig` CR does not allocate anything. The VPC is. The VPC is in the AWS API. We have to work against it. Everything else is designed to fail. I do not care where this work is being done but it has to be done. When you do not want to do this in `cluster-operator` this is fine with me. Ignoring this very point is not fine with me.

Reply:
I would like to end this discussion now. If there is need to talk more about this matter we should schedule a session where we get together to come up with a solution and then Timo should be in there as well.

Reply:
As a final proposal before scheduling discussion session:

I understand @xh3b4sd point on cloud provider specific present state that can diverge from CRs. This is clearly a bug, but it's something we still need to deal with on daily basis.

I do not agree on that `cluster-operator` has any provider specific logic except CR management. One clear boundary there is lack of cloud provider specific API usage outside of e2e tests.

Therefore I propose that IPAM functionality is to be implemented in provider specific operators such as `aws-operator` with shared allocation algorithm from `ipam` library. This way we separate provider specific state management from abstract allocation strategy and implement state management where it's implemented for all other parts as well.

If you agree with this, I'll proceed on implementing IPAM on `aws-operator`. If you disagree, I'll schedule a discussion session.